src/search/migration/migration-manager.ts

```diff
@@ -0,0 +1,138 @@
+import { browser, Storage } from 'webextension-polyfill-ts'
+import whenAllSettled from 'when-all-settled'
+
+import exportOldPages, { ExportParams } from '../search-index-old/export'
+import importNewPage from '../search-index-new/import'
+
+export interface Props {
+    concurrency: number
+    onComplete: () => void
+}
+
+export class MigrationManager {
+    public static PROGRESS_STORAGE_KEY = 'migration-progress'
+    public static FINISHED_STATE = ''
+    public static DEF_PARAMS: ExportParams = {
+        chunkSize: 10,
+        startKey: 'page/',
+        endKey: 'page/\uffff',
+    }
+
+    private static persistProgressState = (
+        key = MigrationManager.FINISHED_STATE,
+    ) =>
+        browser.storage.local.set({
+            [MigrationManager.PROGRESS_STORAGE_KEY]: key,
+        })
+
+    /**
+     * Acts as a flag to stop iteration over old pages if ever set.
+     */
+    private isCancelled = false
+
+    /**
+     * Acts as a progress state, holding the page key to start iterating from.
+     */
+    private currKey: string
+
+    private concurrency: number
+    private onComplete: () => void
+
+    constructor({
+        concurrency = MigrationManager.DEF_PARAMS.chunkSize,
+        onComplete = () => false,
+    }: Partial<Props>) {
+        this.concurrency = concurrency
+        this.onComplete = onComplete
+
+        this.rehydrateProgressState()
+    }
+
+    public get isFinished() {
+        return this.currKey === MigrationManager.FINISHED_STATE
+    }
+
+    /**
+     * Attempts to rehydrate previously persisted progress state, or sets default.
+     */
+    private async rehydrateProgressState() {
+        try {
+            const {
+                [MigrationManager.PROGRESS_STORAGE_KEY]: storedKey,
+            } = await browser.storage.local.get({
+                [MigrationManager.PROGRESS_STORAGE_KEY]:
+                    MigrationManager.DEF_PARAMS.startKey,
+            })
+
+            this.currKey = storedKey
+        } catch (err) {
+            this.currKey = MigrationManager.DEF_PARAMS.startKey
+        }
+    }
+
+    /**
+     * Always throws an Error to signify interruption to caller.
+     * Saves local and persisted progress states.
+     */
+    private async handleInterruption() {
+        this.isCancelled = false
+        await MigrationManager.persistProgressState(this.currKey)
+
+        throw new Error()
```

Comment:
Could this be an Error subclass with a special property (.type, .paused, etc.) to more easily handle it differently than other errors?

Reply:
Great suggestion!